### PR TITLE
[SPARK-27186][CORE] optimize sortshuffle write

### PR DESCRIPTION
**Pull request by wangjiaochun**
## What changes were proposed in this pull request?
If the SortShuffleWriter.write method records is empty, it should be return directly and no need to keep running，refer to the process of BypassMergeSortShuffleWriter.write. Here are the benefits of no need to create Instance ExternalSorter and tmp file. 
## How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by 10110346** [comment]
This change seems reasonable.

**Comment by srowen** [comment]
Shouldn't the underlying Iterator do this? duplicating the code doesn't seem ideal.

**Comment by wangjiaochun** [comment]
The underlying Iterator do nothing.
so I think it's worth revising. first, ExternalSorter-instantiated is expensive operation, such as there is need to 
estimate the size of all the collection of the Instances. Also create tmp file is needless.

**Comment by srowen** [comment]
I see, it's not a special Iterator implementation. Is it possible to refactor this code block, which looks like appears elsewhere, into any reasonable common function?

**Comment by wangjiaochun** [comment]
Sorry，I have no good idea about your suggestion。

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala [hunk 1]
```diff
@@ -49,6 +49,14 @@ private[spark] class SortShuffleWriter[K, V, C](
 
   /** Write a bunch of records to this task's output */
   override def write(records: Iterator[Product2[K, V]]): Unit = {
+
+    if (!records.hasNext) {
+      val partitionLengths = new Array[Long](dep.partitioner.numPartitions)
+      shuffleBlockResolver.writeIndexFileAndCommit(dep.shuffleId, mapId, partitionLengths, null)
+      mapStatus = MapStatus(blockManager.shuffleServerId, partitionLengths)
+      return;
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: Nit: semi colon must be removed

**Review by attilapiros — COMMENTED** [review]
This change really makes sense based on the `BypassMergeSortShuffleWriter.write()` and as `mapStatus` is only `var` which must be set (as it is only which is used latter outside of this class by passing it to the caller of the `stop()` method as part of the return value).

But `SortShuffleManagerSuite` does not test the `write()` method at all. I have checked by adding a `logInfo()` line and executed all the tests of `org.apache.spark.shuffle.sort.*` but that log line cannot be found in `unit-tests.log`.

So I think it would be so great if you could add some tests where this case is tested too.

**Comment by attilapiros** [comment]
@srowen I think it is very hard to refactor this 3-4 lines into a common method as for `SortShuffleWriter.write` the new method should write the `mapStatus` member (which can be solved by passing this `mapStatus` as the return value of the new method) but for `BypassMergeSortShuffleWriter.write` there is two members which should be set (`mapStatus` and `partitionLengths`).

**Comment by attilapiros** [comment]
@wangjiaochun  as I see I am preaching to the choir and you are already started to create a test in #24080.

**Comment by attilapiros** [comment]
Ok to test

**Comment by attilapiros** [comment]
Ok to test

**Comment by attilapiros** [comment]
OK to test

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by JoshRosen** [comment]
Is this actually a performance issue in practice? I see that the JIRA (https://issues.apache.org/jira/browse/SPARK-27186) was closed as "Won't Fix". If this proposed change doesn't give us significant benefit then I'm in favor of skipping it, mostly from a change/risk-management perspective.

**Pull request closed without merging**